src/keywrap/jwe.rs

```diff
@@ -0,0 +1,230 @@
+use anyhow::{anyhow, Result};
+use std::collections::HashMap;
+
+use crate::config::DecryptConfig;
+use crate::config::EncryptConfig;
+use crate::keywrap::KeyWrapper;
+use josekit::jwk::{Jwk, KeyAlg, KeyFormat, KeyInfo};
+
+use josekit::jwe::{
+    deserialize_json, serialize_general_json, JweDecrypter, JweEncrypter, JweHeader, JweHeaderSet,
+    ECDH_ES_A256KW, RSA_OAEP,
+};
+
+/// A Jwe keywrapper
+#[derive(Debug)]
+pub struct JweKeyWrapper {}
+
+// Get the encrypter from public key
+fn get_encrypter(pubkey: &[u8]) -> Result<Box<dyn JweEncrypter>> {
+    let key_info = match KeyInfo::detect(&pubkey) {
+        Some(i) => i,
+        None => return Err(anyhow!("failed to detect public key info")),
+    };
+
+    if !key_info.is_public_key() {
+        return Err(anyhow!("expect public key, found private key"));
+    }
+
+    let alg = match key_info.alg().unwrap() {
+        KeyAlg::Ec { curve: Some(_) } => "ECDH_ES_A256KW",
+        _ => "RSA_OAEP",
+    };
+
+    if alg == "RSA_OAEP" {
+        let rsa_encrypter = match key_info.format() {
+            KeyFormat::Der { raw: _ } => RSA_OAEP.encrypter_from_der(pubkey)?,
+            KeyFormat::Pem { traditional: _ } => RSA_OAEP.encrypter_from_pem(pubkey)?,
+            KeyFormat::Jwk => RSA_OAEP.encrypter_from_jwk(&Jwk::from_bytes(pubkey)?)?,
+        };
+
+        Ok(rsa_encrypter.box_clone())
+    } else {
+        let ecdh_encrypter = match key_info.format() {
+            KeyFormat::Der { raw: _ } => ECDH_ES_A256KW.encrypter_from_der(pubkey)?,
+            KeyFormat::Pem { traditional: _ } => ECDH_ES_A256KW.encrypter_from_pem(pubkey)?,
+            KeyFormat::Jwk => ECDH_ES_A256KW.encrypter_from_jwk(&Jwk::from_bytes(pubkey)?)?,
+        };
```

Comment:
Why can't we do:

```Rust
match key_info.alg().map_err()? {
    KeyAlg::Ec { curve: Some(_) } => {// Prepare and return  the ecdh encrypter}
    _ => {// Prepare and return the rsa encrypter}
}
```

And avoid using the temporary variable `alg` ?

Reply:
Thanks, key_info.alg() is a Option value, I'll match it directly, now the implementation will much clean.